python/pyspark/pandas/data_type_ops/num_ops.py

```diff
@@ -395,6 +401,17 @@ def restore(self, col: pd.Series) -> pd.Series:
         """Restore column when to_pandas."""
         return col.astype(self.dtype)
 
+    def astype(self, index_ops: IndexOpsLike, dtype: Union[str, type, Dtype]) -> IndexOpsLike:
+        dtype, spark_type = pandas_on_spark_type(dtype)
+        if index_ops.hasnans:
+            if is_integer_dtype(dtype) and not isinstance(dtype, extension_dtypes):
+                raise ValueError(
+                    "Cannot convert %s with missing values to integer" % self.pretty_name
+                )
+            elif is_bool_dtype(dtype) and not isinstance(dtype, extension_dtypes):
+                raise ValueError("Cannot convert %s with missing values to bool" % self.pretty_name)
```

Comment:
Could you delay `index_ops.hasnans` to avoid unnecessary Spark jobs?

```py
if is_integer_dtype(dtype) and not isinstance(dtype, extension_dtypes):
    if index_ops.hasnans:
        raise ...
elif is_bool_dtype(dtype) and not isinstance(dtype, extension_dtypes):
    if index_ops.hasnans:
        raise ...
````

Reply:
Good idea! Updated.